### DISCUSSION TITLE
Envoy Mesh, Java Agent & Fluentd

### DISCUSSION THREAD

**Discussion by andrewgkew** [apache/skywalking#7566] [Q&A]
I am looking at configuring SW to communicate with an Envoy Mesh, and because the mesh installs their own agent for each application I am assuming I dont need have the Java Agent setup?

But then how will the log collection work? If I am not using the Java agent I wont using logj42 gRPC client appender and satellite to forward the logs so I assume I can use fluentd instead, but looking at the fluentd configuration it is dependent on the log format to be specific. 

Should I ensure that I setup my log4j2 appender with the right format so that it matches this express

```
expression /^\[SW_CTX:\s*(?<SW_CTX>[^ ]*]*)\] \[(?<level>[^ ]*)\] (?<logtime>[^\]]*) \[(?<thread>[^ ]*)\] (?<logger>[^\]]*):(?<line>[^\]]*) - (?<body>[^\]]*)$/
```
Any docs on this log format?

Do you recommend install fluentd as a daemonSet?

It looks like with fluentd we can push the logs through kafka to elastic search too, so if we are using ES as the backend for SW could we not us that or does OAP need to fetch the data from kafka instead?

**Comment by kezhenxu94** [comment]
> I am looking at configuring SW to communicate with an Envoy Mesh, and because the mesh installs their own agent for each application I am assuming I dont need have the Java Agent setup?

The 2 agents are of different functionalities and they can co-exist, feel free to install Java agent if you need our functionalities.

> But then how will the log collection work? If I am not using the Java agent I wont using logj42 gRPC client appender and satellite to forward the logs so I assume I can use fluentd instead, but looking at the fluentd configuration it is dependent on the log format to be specific.

For Envoy mesh logs, it consists of 2 parts, Envoy access logs and user application logs.

- Envoy access logs: (assuming you're using Istio) you can get metrics, topology and collect envoy access logs themself, see https://skywalking.apache.org/blog/2020-12-03-obs-service-mesh-with-sw-and-als/
- User application logs: you can use fluentd / fluent bit, see our test cases https://github.com/apache/skywalking/blob/master/test/e2e/e2e-test/docker/log/docker-compose.fluentbit.yml and https://github.com/apache/skywalking/tree/master/test/e2e/e2e-test/docker/log/fluent-bit, there are some customization to reformat the logs into SkyWalking acceptable format.

> Should I ensure that I setup my log4j2 appender with the right format so that it matches this express
> 
> ```
> expression /^\[SW_CTX:\s*(?<SW_CTX>[^ ]*]*)\] \[(?<level>[^ ]*)\] (?<logtime>[^\]]*) \[(?<thread>[^ ]*)\] (?<logger>[^\]]*):(?<line>[^\]]*) - (?<body>[^\]]*)$/
> ```

The doc format depends on your applications, just modify the expression to match your format.

> Any docs on this log format?
> 
> Do you recommend install fluentd as a daemonSet?

I recommend fluent-bit (daemonset) as it's specifically optimized to be used in containers.

> It looks like with fluentd we can push the logs through kafka to elastic search too, so if we are using ES as the backend for SW could we not us that or does OAP need to fetch the data from kafka instead?

Yes, fluentd doesn't support gRPC or http, and SkyWalking only supports gRPC, HTTP, Kafka, so you can only choose kafka if you insist using FluentD

**Reply by andrewgkew** [reply]
> > I am looking at configuring SW to communicate with an Envoy Mesh, and because the mesh installs their own agent for each application I am assuming I dont need have the Java Agent setup?
> 
> The 2 agents are of different functionalities and they can co-exist, feel free to install Java agent if you need our functionalities.
> 

@kezhenxu94 - can you explain how this would work.

If I installed Istio and configured the mesh to talk to Skywalking, I would get the APM data and Topology from the mesh ALS (I think) which is great.
But then if I also add the Java agent to my pods they would be sending the same info to Skywalking as well? I realize if I wanted to use functionality from the java agent like for logging then maybe I need both setup and configured.

My questions:
1. Would there be any issue configuring both an agent and mesh at the same time?
2. Am I right in saying they are doing the same thing? i.e. sending the same data across to OAP
3. If I had, mesh, an agent and satellite all configured together would that be ok?

**Comment by wu-sheng** [comment]
Do you want to switch to Service Mesh observability solution? That is totally different. It analyzes Envoy logs.

Read this
https://skywalking.apache.org/docs/main/v8.7.0/en/setup/envoy/als_setting/

We have DP and CP monitoring without tracing agents. If you have agent deployed, you have more data already.

**Reply by andrewgkew** [reply]
@wu-sheng - thanks for this. I have been able to setup skywalking with Istio based on this blog: https://skywalking.apache.org/blog/2020-12-03-obs-service-mesh-with-sw-and-als/ and its all good I get the APM data and trace data into skywalking which is perfect, but what I am confused about is how to:

1. Get data on the Istio tab
2. Get data on the Istio Data Plane tab

Can you help please?

**Reply by andrewgkew** [reply]
Actually I think 1. is the Istio control plane and is based off this page

https://skywalking.apache.org/docs/main/v8.7.0/en/setup/istio/readme/

**Reply by andrewgkew** [reply]
I got the control plane to work by configuring the prometheus exporter with the scrape configs as per the Istio docs using OTEL which is great, now just the data plane.
I cant see any docs around this? Could you point me in the right direction please?

**Reply by wu-sheng** [reply]
https://skywalking.apache.org/docs/main/v8.7.0/en/setup/envoy/metrics_service_setting/

Envoy metrics are DP metrics.

**Reply by andrewgkew** [reply]
Thanks @wu-sheng - I have already set that up with Istio with following

istioctl  manifest install \
    --set meshConfig.enableEnvoyAccessLogService=true \
    --set meshConfig.defaultConfig.envoyAccessLogService.address=skywalking-oap.default.svc.cluster.local:11800

I get the control plane metrics and the APM data from the APIs (via the mesh) but not the DP metrics?

Also when metrics come through for CP there is nothing for memory or CPU? Could that be an cluster permissions issue?

**Reply by wu-sheng** [reply]
I shared with you the link of doc about Envoy metrics, which is collected through metrics service directly. What you shared is Envoy Access Log Service, which are different. Envoy metrics are memory, CPU, etc.

CP metrics are through OpenTelemetry collector and SkyWalking's OC receiver, read this. https://skywalking.apache.org/docs/main/latest/en/setup/istio/readme/

They are read very similar, but those are different these.
1. Envoy Access Log is traffic related metrics.
2. CP is collected through OpenTelemetry collector, and forward to SkyWalking
3. DP metrics are reported by Envoy Metrics Service

**Reply by andrewgkew** [reply]
Hi @wu-sheng - sorry yes you are right they are so similiar I think I over looked that, so I need to update istio with

  --set meshConfig.defaultConfig.envoyMetricsService.address=skywalking-oap.default.svc.cluster.local:11800
  --set 'meshConfig.defaultConfig.proxyStatsMatcher.inclusionRegexps[0]=.*'

I will give that a go, thank you. Setting these parameters should then give me the DP metrics and the CP CPU and memory metrics I assume?

**Reply by wu-sheng** [reply]
I can't remember all, but this env(https://skywalking.apache.org/#demo) will show all we expected. Mesh and language agents are deployed there.

**Reply by andrewgkew** [reply]
Thanks yes I have looked at that demo. What would be super helpful is if what was used to setup that demo was in a repo as like an example this is how you setup everything in skywalking with an example app (guess the booking app) and almost use it as a template to give everything up and running as a demo template and then any tweaks wanted can then be made off this template.

**Reply by wu-sheng** [reply]
@kezhenxu94 It seems a good idea. I think we have one private? We could consider never release `skywalking-service-mesh-example` repo.(Not urgent, but we should consider)

**Reply by kezhenxu94** [reply]
Do we want to create a new repo to put those deploy scripts?

**Reply by wu-sheng** [reply]
I have thought much. We could have more discussion about this later.

**Reply by andrewgkew** [reply]
@wu-sheng - I followed your advice and now have the graphs for both my CP and DP on Istio thank you.

A couple quick points though:

1. One of the Istio manifest settings suggested in your docs was
--set 'meshConfig.defaultConfig.ProxyStatsMatcher.inclusionRegexps[0]=.*'
But when apply that I get the following error

Run the command with the --force flag if you want to ignore the validation error and proceed.
Error: failed to install manifests: failed to unmarshall mesh config: unknown field "ProxyStatsMatcher" in v1alpha1.ProxyConfig

But looking at the Istio docs that field seems to be there, have you come across this before?

2. In my DP graphs, only the 1st line of graphs show any data, the ones below it (Membership history, Upstream CX Active... all the way to Upstream CX None Healthy Increase (pm) are all greys out, not no data greyed out so arent being used is that expected?

3. In the CP graphs, they are all ungreyed but for member and CPU it shows as zero, I would expect there to be something at least?

**Reply by wu-sheng** [reply]
@wankai123 I think you know more about this than me.

**Reply by kezhenxu94** [reply]
> 
> 1. One of the Istio manifest settings suggested in your docs was
> 
> --set 'meshConfig.defaultConfig.ProxyStatsMatcher.inclusionRegexps[0]=.*'
> 
> But when apply that I get the following error
> 
> 
> 
> Run the command with the --force flag if you want to ignore the validation error and proceed.
> 
> Error: failed to install manifests: failed to unmarshall mesh config: unknown field "ProxyStatsMatcher" in v1alpha1.ProxyConfig
> 
> 
> 
> But looking at the Istio docs that field seems to be there, have you come across this before?
> 

This requires istio version >= 1.8.x, make sure you are looking up the right doc and using the needed version.

**Reply by andrewgkew** [reply]
Morning @kezhenxu94 

I tried that just installed Istio 1.11.2 and got the same error

```
 ~/istio$ istioctl  manifest install \
 >     --set meshConfig.enableEnvoyAccessLogService=true \
 >     --set meshConfig.defaultConfig.envoyAccessLogService.address=skywalking-oap.skywalking.svc.cluster.local:11800 \
 >     --set meshConfig.defaultConfig.envoyMetricsService.address=skywalking-oap.skywalking.svc.cluster.local:11800 \
 >     --set 'meshConfig.defaultConfig.ProxyStatsMatcher.inclusionRegexps[0]=.*'
 Detected that your cluster does not support third party JWT authentication. Falling back to less secure first party JWT. See 
 https://istio.io/v1.11/docs/ops/best-practices/security/#configure-third-party-service-account-tokens for details.
 Run the command with the --force flag if you want to ignore the validation error and proceed.
 Error: generate config: failed to unmarshall mesh config: unknown field "ProxyStatsMatcher" in v1alpha1.ProxyConfig
```

**Reply by kezhenxu94** [reply]
s/ProxyStatsMatcher/proxyStatsMatcher

**Reply by andrewgkew** [reply]
Sorry that was my fault, I was trying to fix it (before the version) and thought it might have been a spelling mistake, let me give this a go. thanks

**Reply by andrewgkew** [reply]
Update that and it worked, thanks @kezhenxu94
